cardano-api/src/Cardano/Api/Environment.hs

```diff
@@ -1,44 +1,87 @@
 {-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE LambdaCase #-}
 {-# LANGUAGE OverloadedStrings #-}
 
 module Cardano.Api.Environment
-  ( EnvSocketError(..)
+  ( EnvSocketLookupError(..)
   , SocketPath(..)
   , readEnvSocketPath
   , renderEnvSocketError
+  , EnvNetworkIdError(..)
+  , readEnvNetworkId
+  , parseNetworkId
   ) where
 
 import           Prelude
 
+import           Control.Monad
 import           Data.Aeson
 import           Data.Text (Text)
 import qualified Data.Text as Text
+import           Text.ParserCombinators.ReadP
+
 import           System.Environment (lookupEnv)
 
+import           Cardano.Api.Error
+import           Cardano.Api.NetworkId
 import           Cardano.Api.Utils (textShow)
 
 newtype SocketPath
   = SocketPath { unSocketPath :: FilePath }
   deriving (FromJSON, Show, Eq, Ord)
 
-newtype EnvSocketError = CliEnvVarLookup Text deriving Show
+newtype EnvSocketLookupError = EnvSocketLookupError Text deriving Show
 
-renderEnvSocketError :: EnvSocketError -> Text
+renderEnvSocketError :: EnvSocketLookupError -> Text
 renderEnvSocketError err =
   case err of
-    CliEnvVarLookup txt ->
+    EnvSocketLookupError txt ->
       "Error while looking up environment variable: CARDANO_NODE_SOCKET_PATH " <> " Error: " <> textShow txt
 
 -- | Read the node socket path from the environment.
 -- Fails if the environment variable is not set.
-readEnvSocketPath :: IO (Either EnvSocketError SocketPath)
+readEnvSocketPath :: IO (Either EnvSocketLookupError SocketPath)
 readEnvSocketPath = do
     mEnvName <- lookupEnv envName
     case mEnvName of
       Just sPath ->
         return . Right $ SocketPath sPath
       Nothing ->
-        return . Left $ CliEnvVarLookup (Text.pack envName)
+        return . Left $ EnvSocketLookupError (Text.pack envName)
   where
     envName :: String
     envName = "CARDANO_NODE_SOCKET_PATH"
+
+networkIdEnvName :: String
+networkIdEnvName = "CARDANO_NODE_NETWORK_ID"
+
+readEnvNetworkId :: IO (Either EnvNetworkIdError NetworkId)
+readEnvNetworkId
+  = lookupEnv networkIdEnvName >>= \case
+    Nothing -> return $ Left EnvVarNotSetError
+    Just val -> case parseNetworkId val of
+      Right nid -> return $ Right nid
+      Left err -> return $ Left $ ParsingEnvVarError err
+
+data EnvNetworkIdError
+  = EnvVarNotSetError
+  | ParsingEnvVarError String
+  deriving Show
+
+instance Error EnvNetworkIdError where
+  displayError = \case
+    EnvVarNotSetError -> "Error while looking up environment variable: "<> networkIdEnvName
+    ParsingEnvVarError msg -> "Error while parsing " <> networkIdEnvName <> ": " <> msg
+
+parseNetworkId :: String -> Either String NetworkId
```

Comment:
What do you think of this: https://github.com/input-output-hk/cardano-node/commit/1f4659fdc37c944bf611d562cab0451fbbc2387b

We can then re-export `decimal` to be used in `cardano-cli` and we also continue to use `parsec` which provides better errors than `readS_to_P reads`

Reply:
Thanks for the hints on `parsec` !